### PR TITLE
Add put to groupcache

### PR DESCRIPTION
**Pull request by dgdonut**
Builds on the Getter/Get interfaces to implement Putter/Put and support them in Peers and HTTPPools. Adds basic tests, but duplicated "dedup" and other logic is not retested.

Puts act like Gets in that they do a local cache check (no-op), and then determine the correct Peer for the key being written. The local machine will then either Put locally or via the correct peer based on the consistent hash of the key. The machine that processes the Put will add the data to its main cache. In a stable peer environment, this means any Get after a Put should be cached (barring exceptions) because the canonical peer cached the data already.

Will test in Scoot to verify it works in desired scenarios before landing.

There is no CI here, pre-submit check is manual fmt, vet, & test -race.

### CODE REVIEW

File: groupcache.go [hunk 1]
```diff
@@ -232,10 +269,32 @@ func (g *Group) Get(ctx Context, key string, dest Sink) error {
...
+	if cacheHit {
+		g.Stats.CacheHits.Add(1)
+		return nil
+	}
+
+	// Optimization to avoid double unmarshalling or copying: keep
+	// track of whether the dest was already populated. One caller
+	// (if local) will set this; the losers will not. The common
+	// case will likely be one caller.
+	destPopulated := false
```
Comment (jasonbs10) [inline groupcache.go]: Is this still meaningful?

File: groupcache.go [hunk 2]
```diff
@@ -298,6 +357,39 @@ func (g *Group) load(ctx Context, key string, dest Sink) (value ByteView, destPo
...
+			}
+			g.Stats.PeerErrors.Add(1)
+		}
+		err = g.putLocally(ctx, key, data)
+		if err != nil {
+			g.Stats.LocalStoreErrs.Add(1)
+			return nil, err
+		}
+		g.Stats.LocalStores.Add(1)
+		destPopulated = true // only one caller of load gets this return value
```
Comment (jasonbs10) [inline groupcache.go]: Update comment.

File: groupcache.go [hunk 3]
```diff
@@ -326,6 +418,31 @@ func (g *Group) getFromPeer(ctx Context, peer ProtoGetter, key string) (ByteView
...
+	res := &pb.PutResponse{}
+	err := peer.Put(ctx, req, res)
+	if err != nil {
+		return err
+	}
+	value := ByteView{b: data}
+	// TODO(bradfitz): use res.MinuteQps or something smart to
+	// conditionally populate hotCache.  For now just do it some
+	// percentage of the time.
+	if rand.Intn(10) == 0 {
```
Comment (jasonbs10) [inline groupcache.go:435]: Can you do some napkin math and determine if we should be more aggressive here and in the equivalent 'get' above?

File: http.go [hunk 4]
```diff
@@ -161,26 +161,50 @@ func (p *HTTPPool) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		body, err := proto.Marshal(&pb.GetResponse{Value: value})
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "application/x-protobuf")
+		w.Write(body)
+	case "POST":
+		buf := new(bytes.Buffer)
+		buf.ReadFrom(r.Body)
```
Comment (jasonbs10) [inline http.go:185]: Do you need to close body?

File: groupcache.go [hunk 5]
```diff
@@ -54,6 +56,30 @@ func (f GetterFunc) Get(ctx Context, key string, dest Sink) error {
...
+	// Data cannot be invalidated - it is assumed that
+	// putting any data with a preexisting key can be
+	// interpreted as a no-op.
+	Put(ctx Context, key string, data []byte) error
+}
+
+// A PutterFunc implements Putter with a function.
+type PutterFunc func(ctx Context, key string, data []byte) error
+
+func (f PutterFunc) Put(ctx Context, key string, data []byte) error {
```
Comment (jasonbs10) [inline groupcache.go:73]: It's been a while, the difference between Put and store is unclear here (as is Get and load) and should be made explicit in the comments.

**Review by jasonbs10 — APPROVED** [review]
Since this pretty much mirrors the get/load functions, I'll trust that it's doing the right thing.

**Comment by jasonbs10** [comment]
Do we need continuous integration set up if you'll be working on this?

File: groupcache.go (hunk 1 above)
Comment (dgdonut) [reply groupcache.go]: nope, seems useless, removing

File: groupcache.go (hunk 2 above)
Comment (dgdonut) [reply groupcache.go]: removing

File: http.go (hunk 4 above)
Comment (dgdonut) [reply http.go:185]: "The Server will close the request body. The ServeHTTP Handler does not need to." per docs

File: groupcache.go (hunk 5 above)
Comment (dgdonut) [reply groupcache.go:73]: adding comment and reordering functions

**Comment by dgdonut** [comment]
Taking a look at a basic travis setup.

For the hot cache 10% figure, I don't think we'd see much benefit, but we could set it to 100% without any negative impacts. I'm interested in using the Context params to make this deterministic for certain use cases like "store in cache only", or "try to propagate to cache on all peers", but it's not clear yet if we need to do this (and requires more scoot interface modification).

**Comment by dgdonut** [comment]
Travis isn't set up for twitter-forks, so I'm going to skip that.

**Pull request merged**